static/compiler.js

```diff
@@ -437,15 +639,177 @@ function Compiler(domRoot, origFilters, windowLocalPrefix, onChangeCallback, lan
 
     function setFilters(f) {
         filters_ = $.extend({}, f);
-        onChange();
-        updateCompilerAndButtons();
+        slots.forEach(function (slot) {
+            onParamChange(slot);
+            updateCompilerAndButtons(slot);
+        });
     }
 
     function setEditorHeight(height) {
-        const MinHeight = 100;
+        const MinHeight = 80;
         if (height < MinHeight) height = MinHeight;
         cppEditor.setSize(null, height);
-        asmCodeMirror.setSize(null, height);
+        for (var i = 0; i < slots.length; i++) {
+            slots[i].asmCodeMirror.setSize(null, height);
+        }
+    }
+
+    // External wrapper used by gcc.js only
+    function refreshSlot(slot) {
+        onCompilerChange(slot);
+    }
+    
+    // Auxiliary function to slot_use_DOM
+    function generate_change_callback(slot) {
+        return function callback() {
+            onParamChange(slot);
+        }
+    }
+
+    // Function to call each time a slot is added to the page.
+    // This function requires that the slot's DOM object already exists.
+    function slot_use_DOM(slot) {
+        slot.asmCodeMirror = CodeMirror.fromTextArea(
+        domRoot.find("#slot"+slot.id+" .asm textarea")[0], {
+            lineNumbers: true,
+            mode: "text/x-asm",
+            readOnly: true,
+            gutters: ['CodeMirror-linenumbers'],
+            lineWrapping: true
+        });
+        // handle compiler option (slot specific) such as '-O1'
+        domRoot.find('#slot'+slot.id+' .compiler_options').change(
+            generate_change_callback(slot)).keyup(
+            generate_change_callback(slot));
+
+        setLeaderSlotMenu();
+        if (slots.length == 1) {
+            // "force" menu update if this is the first slot added
+            var leaderSlotMenuNode = domRoot.find('#commonParams .leaderSlot');
+            leaderSlotMenuNode.text('leader slot : '+0);
+        }
+    }
+
+    function get_slots_ids() {
+        var ids = [];
+        for (var i = 0; i < slots.length; i++) {
+            ids.push(slots[i].id);
+        }
+        return ids;
+    }
+
+    function slot_ctor(optionalId) {
+        var newSlot = new Slot();
+        if (optionalId) {
+            newSlot.id = optionalId;
+        } else {
+            newSlot.id = get_available_id(slots);
+        }
+        slots.push(newSlot);
+        setSetting('slotsId',JSON.stringify(get_slots_ids()));
+        return newSlot;
+    }
+    
+    // Array Remove - By John Resig (MIT Licensed)
```

Comment:
Isn't package.json a npm package ? This code runs in the browser. 
Should I download https://github.com/jashkenas/underscore in static/underscore (no as submodule) ?